### PR TITLE
Fix/egl buffer segfault

### PR DESCRIPTION
**Pull request by cmeissl**
This PR fixes some segfaults for egl clients that have disconnected

- [9b8b86a](https://github.com/Smithay/smithay/commit/9b8b86ad43b2bf0bbed65312fcb3efd1a2d9245d) updates `wayland-rs` to `0.30.0-beta10` and makes use of `io_lifetimes`
- [05b1606](https://github.com/Smithay/smithay/commit/05b1606eb534b593789bb6fe263e77402f565bdd) fixes a race in the renderer accessing a destroyed EGL Buffer

### CODE REVIEW

**Comment by cmeissl** [comment]
@vberger the build breaks because I also wanted to use `OwnedFd` for the `Dmabuf` handling in smithay and just used the re-export from `wayland_server`. What do you think about that issue? Should we add `io_lifetimes` as a dependecy to smithay instead of relying on the `wayland_server` re-export?

**Comment by i509VCB** [comment]
I think smithay depending on io-lifetimes makes sense. We don't always have wayland-server (Wayland frontend feature)

File: src/lib.rs [hunk 1]
```diff
@@ -49,6 +49,8 @@
 //! `log` crate. If not, the logs will discarded. This cargo feature is part of the default set of
 //! features of Smithay.
 
+#[doc(hidden)]
+pub extern crate io_lifetimes;
```
Comment (i509VCB) [inline src/lib.rs:53]: Why do we need an extern crate? This isn't 2015

Comment (cmeissl) [reply src/lib.rs:53]: I know, I just aligned it with `nix` just below, but we can also change both?

File: src/backend/egl/display.rs [hunk 2]
```diff
@@ -906,6 +906,10 @@ impl EGLBufferReader {
         &self,
         buffer: &WlBuffer,
     ) -> ::std::result::Result<EGLBuffer, BufferAccessError> {
+        if !buffer.is_alive() {
+            return Err(BufferAccessError::NotManaged(EGLError::BadAccess));
```
Comment (cmeissl) [inline src/backend/egl/display.rs:910]: @Drakulix @i509VCB not sure if this is the best suited error, maybe something like `BufferAccessError::Destroyed` would be a better fit?

File: src/wayland/data_device/mod.rs [hunk 3]
```diff
@@ -114,7 +115,7 @@ pub trait DataDeviceHandler: Sized + ClientDndGrabHandler + ServerDndGrabHandler
     ///
     /// * `mime_type` - the requested mime type
     /// * `fd` - the fd to write into
-    fn send_selection(&mut self, mime_type: String, fd: RawFd) {}
+    fn send_selection(&mut self, mime_type: String, fd: OwnedFd) {}
```
Comment (cmeissl) [inline src/wayland/data_device/mod.rs:118]: @vberger is this correct to just close the fd by dropping it here

File: src/wayland/data_device/mod.rs [hunk 4]
```diff
@@ -160,7 +161,7 @@ pub trait ServerDndGrabHandler {
     ///
     /// * `mime_type` - The requested mime type
     /// * `fd` - The FD to write into
-    fn send(&mut self, mime_type: String, fd: RawFd) {}
+    fn send(&mut self, mime_type: String, fd: OwnedFd) {}
```
Comment (cmeissl) [inline src/wayland/data_device/mod.rs:164]: @vberger same here

File: src/wayland/data_device/mod.rs (hunk 3 above)
Comment (vberger) [reply src/wayland/data_device/mod.rs:118]: Looks right, the function passes FD ownership to the handler, which then decides what to do with it. :+1:

File: src/backend/egl/display.rs (hunk 2 above)
Comment (vberger) [reply src/backend/egl/display.rs:910]: `BufferAccessError::Destroyed` sounds better than `BufferAccessError::NotManaged` yeah

File: src/lib.rs (hunk 1 above)
Comment (vberger) [reply src/lib.rs:53]: I think `nix` and `io_lifetimes` should just be `pub use`-ed in the `reexports` module.

Comment (cmeissl) [reply src/lib.rs]: I moved `io_lifetimes` to the `reexports` and removed the `pub extern crate nix`.
The `nix` crate already was in the `reexports`, but behind some feature gates. I made it available unconditonally as `nix` is not an optional crate anyway.

File: src/backend/egl/display.rs (hunk 2 above)
Comment (cmeissl) [reply src/backend/egl/display.rs]: Added `BufferAccessError:Destroyed` and updated `egl_buffer_contents`

**Review by Drakulix — APPROVED** [review]

**Pull request merged**